### PR TITLE
Support multiple blockquotes & enable range grouping

### PR DESCRIPTION
**Pull request by robertKozik**
### Related Issues

GH_LINK

### CODE REVIEW

File: android/src/main/java/com/expensify/livemarkdown/MarkdownBlockquoteSpan.java [hunk 1]
```diff
@@ -36,10 +38,14 @@ public void drawLeadingMargin(Canvas c, Paint p, int x, int dir, int top, int ba
 
     p.setStyle(Paint.Style.FILL);
     p.setColor(borderColor);
+    for(int stripe = 0; stripe < nestLevel; stripe++) {
```
Comment (tomekzaw) [inline android/src/main/java/com/expensify/livemarkdown/MarkdownBlockquoteSpan.java:41]: Let's add a space between `for` and `(` and maybe rename the variable?
```suggestion
    for (int level = 0; level < nestLevel; level++) {
```

File: android/src/main/java/com/expensify/livemarkdown/MarkdownBlockquoteSpan.java [hunk 2]
```diff
@@ -36,10 +38,14 @@ public void drawLeadingMargin(Canvas c, Paint p, int x, int dir, int top, int ba
...
+
+      float shift = (borderWidth + marginLeft + paddingLeft) * stripe;
```
Comment (tomekzaw) [inline android/src/main/java/com/expensify/livemarkdown/MarkdownBlockquoteSpan.java:43]: Let's make it consistent with `getLeadingMargin`.
```suggestion
      float shift = (marginLeft + borderWidth + paddingLeft) * stripe;
```

File: android/src/main/java/com/expensify/livemarkdown/MarkdownUtils.java [hunk 3]
```diff
@@ -91,18 +91,19 @@ public void applyMarkdownFormatting(SpannableStringBuilder ssb) {
     try {
       JSONArray ranges = new JSONArray(output);
       for (int i = 0; i < ranges.length(); i++) {
-        JSONArray range = ranges.getJSONArray(i);
-        String type = range.getString(0);
-        int start = range.getInt(1);
-        int end = start + range.getInt(2);
-        applyRange(ssb, type, start, end);
+        JSONObject range = ranges.getJSONObject(i);
+        String type = range.getString("style");
```
Comment (tomekzaw) [inline android/src/main/java/com/expensify/livemarkdown/MarkdownUtils.java:95]: Let's stick to either `type` or `style`.
```suggestion
        String style = range.getString("style");
```

File: ios/MarkdownLayoutManager.mm [hunk 4]
```diff
@@ -7,9 +7,11 @@ - (void)drawBackgroundForGlyphRange:(NSRange)glyphsToShow atPoint:(CGPoint)origi
 
   [self enumerateLineFragmentsForGlyphRange:glyphsToShow usingBlock:^(CGRect rect, CGRect usedRect, NSTextContainer * _Nonnull textContainer, NSRange glyphRange, BOOL * _Nonnull stop) {
     __block BOOL isBlockquote = NO;
+    __block int currentDepth = 0;
```
Comment (tomekzaw) [inline ios/MarkdownLayoutManager.mm:10]: I believe we can move the declaration of `currentDepth` into `enumerateObjectsUsingBlock:`.
```suggestion
```

File: ios/MarkdownLayoutManager.mm [hunk 5]
```diff
@@ -7,9 +7,11 @@ - (void)drawBackgroundForGlyphRange:(NSRange)glyphsToShow atPoint:(CGPoint)origi
...
     RCTMarkdownUtils *markdownUtils = [self valueForKey:@"markdownUtils"];
-    [markdownUtils.blockquoteRanges enumerateObjectsUsingBlock:^(NSValue *item, NSUInteger idx, BOOL * _Nonnull stop) {
-      NSRange range = [item rangeValue];
+    [markdownUtils.leveledBlockquoteRanges enumerateObjectsUsingBlock:^(NSDictionary *item, NSUInteger idx, BOOL * _Nonnull stop) {
+      NSRange range = [[item valueForKey:@"range"] rangeValue];
+      currentDepth = [[item valueForKey:@"depth"] unsignedIntegerValue];
```
Comment (tomekzaw) [inline ios/MarkdownLayoutManager.mm:14]:
```suggestion
      NSUInteger depth = [[item valueForKey:@"depth"] unsignedIntegerValue];
```

File: ios/MarkdownLayoutManager.mm [hunk 6]
```diff
@@ -25,9 +27,13 @@ - (void)drawBackgroundForGlyphRange:(NSRange)glyphsToShow atPoint:(CGPoint)origi
       CGFloat y = paddingTop + rect.origin.y;
       CGFloat width = markdownUtils.markdownStyle.blockquoteBorderWidth;
       CGFloat height = rect.size.height;
-      CGRect lineRect = CGRectMake(x, y, width, height);
-      [markdownUtils.markdownStyle.blockquoteBorderColor setFill];
-      UIRectFill(lineRect);
+      CGFloat nestShift = paddingLeft + width + markdownUtils.markdownStyle.blockquoteMarginLeft;
```
Comment (tomekzaw) [inline ios/MarkdownLayoutManager.mm:30]: 1. Let's rename the variable from `nestShift` to `shift` to be consistent with iOS.
2. We should also include `markdownUtils.markdownStyle.blockquotePaddingLeft` (blockquote padding) which is a different thing than `markdownUtils.backedTextInputView.textContainerInset.left` (text input padding). Let's assign all intermediate variables and change this line to:
```suggestion
      CGFloat shift = blockquoteMarginLeft + blockquoteBorderWidth + blockquotePaddingLeft;
```

File: types/global.d.ts [hunk 7]
```diff
@@ -1,6 +1,11 @@
 export {};
 
-type Range = [string, number, number]; // style, location, length
+type Range = {
+  style: string;
+  start: number;
+  length: number;
+  depth?: number;
+}; // style, location, length
```
Comment (tomekzaw) [inline types/global.d.ts:8]: Let's remove this comment since it's no longer relevant
```suggestion
};
```

File: types/global.d.ts [hunk 8]
```diff
@@ -1,6 +1,11 @@
 export {};
 
-type Range = [string, number, number]; // style, location, length
+type Range = {
+  style: string;
```
Comment (tomekzaw) [inline types/global.d.ts:4]: Let's use `MarkdownType` (perhaps it should be named `MarkdownStyle`?) here

File: ios/MarkdownLayoutManager.mm [hunk 9]
```diff
@@ -25,9 +27,13 @@ - (void)drawBackgroundForGlyphRange:(NSRange)glyphsToShow atPoint:(CGPoint)origi
...
+
+      for(int strip = 0; strip < currentDepth; strip++) {
+        CGRect lineRect = CGRectMake(x + (strip * nestShift), y, width, height);
```
Comment (tomekzaw) [inline ios/MarkdownLayoutManager.mm:33]: The idea behind these variables (`x`, `y`, `width`, `height`) was that they are calculated above and thus we don't need any calculations in `CGRectMake`. Please move the calculations outside of it.
```suggestion
        CGRect lineRect = CGRectMake(x, y, width, height);
```

File: ios/RCTMarkdownUtils.mm [hunk 10]
```diff
@@ -55,13 +55,14 @@ - (NSAttributedString *)parseMarkdown:(nullable NSAttributedString *)input
...
   [ranges enumerateObjectsUsingBlock:^(id _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
-    NSArray *item = obj;
-    NSString *type = item[0];
-    NSUInteger location = [item[1] unsignedIntegerValue];
-    NSUInteger length = [item[2] unsignedIntegerValue];
+    NSDictionary *item = obj;
+    NSString *type = [item valueForKey:@"style"];
+    NSInteger location = [[item valueForKey:@"start"] unsignedIntegerValue];
+    NSInteger length = [[item valueForKey:@"length"] unsignedIntegerValue];
+      NSInteger depth = [[item valueForKey:@"depth"] unsignedIntegerValue] ?: 1;
```
Comment (tomekzaw) [inline ios/RCTMarkdownUtils.mm:65]: Too many spaces
```suggestion
    NSInteger depth = [[item valueForKey:@"depth"] unsignedIntegerValue] ?: 1;
```

File: ios/RCTMarkdownUtils.mm [hunk 11]
```diff
@@ -105,10 +106,13 @@ - (NSAttributedString *)parseMarkdown:(nullable NSAttributedString *)input
     } else if ([type isEqualToString:@"blockquote"]) {
       CGFloat indent = _markdownStyle.blockquoteMarginLeft + _markdownStyle.blockquoteBorderWidth + _markdownStyle.blockquotePaddingLeft;
       NSMutableParagraphStyle *paragraphStyle = [NSMutableParagraphStyle new];
-      paragraphStyle.firstLineHeadIndent = indent;
-      paragraphStyle.headIndent = indent;
+      paragraphStyle.firstLineHeadIndent = indent * depth;
+      paragraphStyle.headIndent = indent * depth;
```
Comment (tomekzaw) [inline ios/RCTMarkdownUtils.mm:110]: I've noticed that here we were already using `indent` and you have introduced term `shift` for the same thing. Let's unify them.

File: ios/RCTMarkdownUtils.mm [hunk 12]
```diff
@@ -105,10 +106,13 @@ - (NSAttributedString *)parseMarkdown:(nullable NSAttributedString *)input
...
       [attributedString addAttribute:NSParagraphStyleAttributeName value:paragraphStyle range:range];
-      [_blockquoteRanges addObject:[NSValue valueWithRange:range]];
+
+      NSArray *dictValues = @[[NSValue valueWithRange:range], [NSNumber numberWithInt:depth]];
+      NSArray *keys = @[@"range", @"depth"];
+      [_leveledBlockquoteRanges addObject: [NSDictionary dictionaryWithObjects:dictValues forKeys:keys]];
```
Comment (tomekzaw) [inline ios/RCTMarkdownUtils.mm:115]: Let's use the short-hand syntax for creating dictionaries:
```suggestion
      [_leveledBlockquoteRanges addObject:@{
        @"range": @(range),
        @"depth": @(depth)
      }];
```

File: parser/index.ts [hunk 13]
```diff
@@ -3,8 +3,14 @@
...
-type Range = [string, number, number];
-type Token = ['TEXT' | 'HTML', string];
+type Styles = 'bold' | 'italic' | 'strikethrough' | 'mention-here' | 'mention-user' | 'link' | 'code' | 'pre' | 'blockquote' | 'h1' | 'syntax';
+type Range = {
+  style: Styles;
+  start: number;
+  length: number;
+  depth?: number;
+};
+type Token = ['TEXT' | 'HTML', string, number?];
```
Comment (tomekzaw) [inline parser/index.ts:13]: Why do we need `number?` here?

File: parser/index.ts [hunk 14]
```diff
@@ -3,8 +3,14 @@
 import {ExpensiMark} from 'expensify-common/lib/ExpensiMark';
 import _ from 'underscore';
 
-type Range = [string, number, number];
-type Token = ['TEXT' | 'HTML', string];
+type Styles = 'bold' | 'italic' | 'strikethrough' | 'mention-here' | 'mention-user' | 'link' | 'code' | 'pre' | 'blockquote' | 'h1' | 'syntax';
```
Comment (tomekzaw) [inline parser/index.ts:6]:
```suggestion
type MarkdownStyle = 'bold' | 'italic' | 'strikethrough' | 'mention-here' | 'mention-user' | 'link' | 'code' | 'pre' | 'blockquote' | 'h1' | 'syntax';

```

**Review by tomekzaw — CHANGES_REQUESTED** [review]

File: ios/MarkdownLayoutManager.mm [hunk 15]
```diff
@@ -21,13 +23,16 @@ - (void)drawBackgroundForGlyphRange:(NSRange)glyphsToShow atPoint:(CGPoint)origi
...
       CGFloat paddingTop = markdownUtils.backedTextInputView.textContainerInset.top;
-      CGFloat x = paddingLeft + markdownUtils.markdownStyle.blockquoteMarginLeft;
       CGFloat y = paddingTop + rect.origin.y;
       CGFloat width = markdownUtils.markdownStyle.blockquoteBorderWidth;
       CGFloat height = rect.size.height;
-      CGRect lineRect = CGRectMake(x, y, width, height);
-      [markdownUtils.markdownStyle.blockquoteBorderColor setFill];
-      UIRectFill(lineRect);
+      CGFloat shift = markdownUtils.markdownStyle.blockquoteMarginLeft + markdownUtils.markdownStyle.blockquoteBorderWidth + markdownUtils.markdownStyle.blockquotePaddingLeft;
+      for(int level = 0; level < currentDepth; level++) {
```
Comment (tomekzaw) [inline ios/MarkdownLayoutManager.mm:30]:
```suggestion
      for (int level = 0; level < currentDepth; level++) {
```

File: ios/MarkdownLayoutManager.mm [hunk 16]
```diff
@@ -21,13 +23,16 @@ - (void)drawBackgroundForGlyphRange:(NSRange)glyphsToShow atPoint:(CGPoint)origi
...
+        CGFloat x = paddingLeft + markdownUtils.markdownStyle.blockquoteMarginLeft + (level * shift);
```
Comment (tomekzaw) [inline ios/MarkdownLayoutManager.mm:31]: It makes more sense to me with this ordering:
```suggestion
        CGFloat x = paddingLeft + (level * shift) + markdownUtils.markdownStyle.blockquoteMarginLeft;
```

**Review by tomekzaw — CHANGES_REQUESTED** [review]

**Comment by robertKozik** [comment]
closed in favour of #202

**Pull request closed without merging**